drivers/flash/flash_cadence_nand_ll.c

```diff
@@ -0,0 +1,1617 @@
+		}
+	}
+	return ret;
+}
+
+/**
+ * Perform a generic erase operation for a range of blocks in the Cadence NAND controller.
+ *
+ * @param params The Cadence NAND parameters structure.
+ * @param start_block The starting block number for the erase operation.
+ * @param block_count The number of blocks to be erased.
+ * @retval  0 on success or negative error value on failure.
+ */
+static int cdns_nand_gen_erase(struct cadence_nand_params *params, uint32_t start_block,
+			       uint32_t block_count)
+{
+	uint64_t mini_ctrl_cmd = 0;
+	uintptr_t base_address = 0;
+	uint32_t gen_row_address = 0;
+	uint32_t index = 0;
+	int ret = 0;
+
+	for (index = 0; index < block_count; index++) {
+		row_address_set(params, &gen_row_address, (start_block * params->npages_per_block));
+		start_block++;
+		base_address = params->nand_base;
+		mini_ctrl_cmd |= GCMD_TWB_VALUE;
+		mini_ctrl_cmd |= ERASE_ADDR_SIZE;
+		mini_ctrl_cmd |= ((gen_row_address) & (U32_MASK_VAL));
+		mini_ctrl_cmd |= PAGE_ERASE_CMD;
+		ret = cdns_generic_send_cmd(params, mini_ctrl_cmd);
+		if (ret != 0) {
+			return ret;
+		}
+	}
+	return 0;
+}
+#endif
+
+/**
+ * Read data from the Cadence NAND controller into a buffer.
+ *
+ * @param params The Cadence NAND parameters structure.
+ * @param buffer The buffer to store the read data.
+ * @param offset The offset within the NAND to start reading from.
+ * @param size The size of the data to read.
+ * @retval  0 on success or negative error value on failure.
+ */
+int cdns_nand_read(struct cadence_nand_params *params, const void *buffer, uint32_t offset,
+		   uint32_t size)
+{
+	uint32_t start_page_number;
+	uint32_t end_page_number;
+	uint32_t page_count;
+	int ret = 0;
+	uint16_t r_bytes;
+	uint16_t bytes_dif;
+	uint16_t lp_bytes_dif;
+	uint8_t check_page_first = 0;
+	uint8_t check_page_last = 0;
+	uint8_t *first_end_page;
+	uint8_t *last_end_page;
+
+	if (params == NULL) {
+		LOG_ERR("Wrong parameter passed!!");
+		return -EINVAL;
+	}
+
+	if ((size == 0) || (offset >= params->device_size) ||
+	    (((offset + size) - 1) >= params->device_size)) {
+		LOG_ERR("Wrong offset value passed!!");
+		return -EINVAL;
+	}
+
+	start_page_number = offset / (params->page_size);
+	end_page_number = ((offset + size) - 1) / ((params->page_size));
+
+	if ((offset % params->page_size) == 0) {
+		check_page_first = 1;
+	}
+	if (((offset + size) % params->page_size) == 0) {
+		check_page_last = 1;
+	}
+	page_count = end_page_number - start_page_number;
+	page_count++;
+	if ((check_page_last == 1) && (check_page_first == 1)) {
+#if CONFIG_CDNS_NAND_CDMA_MODE
+		ret = cdns_nand_cdma_read(params, start_page_number, (char *)buffer, page_count);
+#elif CONFIG_CDNS_NAND_PIO_MODE
+		ret = cdns_nand_pio_read_write(params, start_page_number, (char *)buffer,
+					       page_count, CDNS_READ);
+#elif CONFIG_CDNS_NAND_GENERIC_MODE
+		ret = cdns_nand_gen_read_write(params, start_page_number, (char *)buffer,
+					       page_count, CDNS_READ);
+#endif
```

Comment:
This blocks keep on repeating.
Wouldn't it look better if function would be defined, maybe inline, that would change depending on the config defined and called here?

Reply:
@de-nordic i created inline function and updated code